### PR TITLE
Faster processing of scheduler queue activations

### PR DESCRIPTION
**Pull request by serxa**
Scheduler queue activations could be as frequent as resource requests and we have to be able to process them as fast as possible. Allocating a `std::function` with lambda and putting it into a deque is not optimal. This PR adds a separate queue for activations. Logically this queue has a common ordering with `events` deque, which is forced by using common `event_id`s and merging on dequeue. To avoid allocation during enqueue of an activation, queue implementation is an intrusive list. Note that we never put the same `ISchedulerNode` object into `EventQueue` twice, so the node can be safely used as an intrusive-list element.

It is a part of https://github.com/ClickHouse/ClickHouse/issues/63213

### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
...

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All NOT Required Checks
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: Integration Tests
- [ ]  Exclude: Stateless tests
- [ ]  Exclude: Stateful tests
- [ ]  Exclude: Performance tests
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with Aarch64
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
---
- [ ]  Do not test
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-clickhouse** [comment]
*This is an automated comment for commit c02cb392bfebeff6cae375eede4990c5fff20fa4 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/64985/c02cb392bfebeff6cae375eede4990c5fff20fa4/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>A Sync</td><td>If it fails, ask a maintainer for help</td><td>⏳ pending</td></tr>
<tbody>
</table>
<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>Checks correctness of the PR's body</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

File: src/Common/Scheduler/ISchedulerNode.h
Comment (alexkats) [inline src/Common/Scheduler/ISchedulerNode.h:215]: Should it be `Task` instead of `std::function<void()>`?

File: src/Common/Scheduler/ISchedulerNode.h
Comment (alexkats) [inline src/Common/Scheduler/ISchedulerNode.h:183]: Might be a good idea to move `EventId` to a higher level, and use it here instead of `UInt64`, but I'll leave it up to you

**Review by alexkats — APPROVED** [review]

**Comment by evillique** [comment]
From the [logs](https://s3.amazonaws.com/clickhouse-builds/PRs/64993/f658456d02c528d005b3f91ea8cf432f7273cc31/binary_tidy/build_log.log) it looks like this PR has broken clang-tidy build in master

**Comment by serxa** [comment]
Let me fix it real quick

**Comment by serxa** [comment]
https://github.com/ClickHouse/ClickHouse/pull/65415

**Comment by serxa** [comment]
I have no idea how it passed "Mergeable Check" with broken build. Maybe build check is now not required?

**Pull request merged**
